src/network/networkd-dhcp-server.c

```diff
@@ -189,10 +190,99 @@ static int link_push_uplink_sip_to_dhcp_server(Link *link, sd_dhcp_server *s) {
         return sd_dhcp_server_set_sip(s, addresses, n_addresses);
 }
 
+static bool dhcp4_server_emit_servers(Link *link) {
+        Link *uplink = NULL;
+        int r;
+
+        assert(link);
+
+        uplink = manager_find_uplink(link->manager, link);
+        if (!uplink) {
+                log_link_debug(link, "Not emitting DNS, NTP and SIP server information on link, couldn't find suitable uplink.");
+                return false;
+        }
+
+        if (link->network->dhcp_server_emit_dns && link->network->n_dhcp_server_dns > 0) {
+                r = sd_dhcp_server_set_dns(link->dhcp_server, link->network->dhcp_server_dns, link->network->n_dhcp_server_dns);
+                if (r < 0) {
+                        log_link_warning_errno(link, r, "Failed to set DNS server for DHCP server, ignoring: %m");
+                        return false;
+                }
+
+                r = link_push_uplink_dns_to_dhcp_server(uplink, link->dhcp_server);
+                if (r < 0) {
+                        log_link_debug(link, "Not emitting DNS server information on link, couldn't find suitable uplink.");
+                        return false;
+                }
+        }
+
+        if (link->network->dhcp_server_emit_ntp && link->network->n_dhcp_server_ntp > 0) {
+                r = sd_dhcp_server_set_ntp(link->dhcp_server, link->network->dhcp_server_ntp, link->network->n_dhcp_server_ntp);
+                if (r < 0) {
+                        log_link_warning_errno(link, r, "Failed to set NTP server for DHCP server, ignoring: %m");
+                        return false;
+                }
+
+                r = link_push_uplink_ntp_to_dhcp_server(uplink, link->dhcp_server);
+                if (r < 0) {
+                        log_link_debug(link, "Not emitting NTP server information on link, couldn't find suitable uplink.");
+                        return false;
+                }
+        }
+
+        if (link->network->dhcp_server_emit_sip && link->network->n_dhcp_server_sip > 0) {
+                r = sd_dhcp_server_set_sip(link->dhcp_server, link->network->dhcp_server_sip, link->network->n_dhcp_server_sip);
+                if (r < 0) {
+                        log_link_warning_errno(link, r, "Failed to set SIP server for DHCP server, ignoring: %m");
+                        return false;
+                }
+
+                r = link_push_uplink_sip_to_dhcp_server(uplink, link->dhcp_server);
+                if (r < 0) {
+                        log_link_debug(link, "Not emitting sip server information on link, couldn't find suitable uplink.");
+                        return false;
+                }
+        }
+
```

Comment:
we should add support for the other server types you added here I guess, one day.

Reply:
yes